### PR TITLE
AP_BattMonitor: added a "battery" backend for liquid fuel flow

### PR DESCRIPTION
**Pull request by tridge**
This is for sensors that give a pulse for each fixed volume of fuel.

Output is:
    - current in Amps maps to in litres/hour
    - consumed mAh is in consumed millilitres
    - fixed 1.0v voltage

Here is an example sensor suitable for this driver:
https://www.btflowmeter.com/en/flow-meter-products/flow-meters-lowflow-flowmeters-low-flow-turbine-flow-meter-fuel-flow-meters-diesel-fuel-flow-meter-watermeters-paddlewheel-flow-meter-oil-flow-meter-oilflowmeter-waterflowmeter-water-meters-turbineflowmeter-oilflowmeter/mini-flowmeter-fch-mini-pp-series-chemical/fch-m-pp-30-lpm-97478169-lc.html

Thanks to RFDesign for sponsoring the development of this driver!

### CODE REVIEW

File: libraries/AP_BattMonitor/AP_BattMonitor_FuelFlow.cpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
+#include <AP_HAL/AP_HAL.h>
+#include <AP_Common/AP_Common.h>
+#include <AP_Math/AP_Math.h>
+#include "AP_BattMonitor.h"
+#include "AP_BattMonitor_FuelFlow.h"
+#include <GCS_MAVLink/GCS.h>
+#include <stdio.h>
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_FuelFlow.cpp]: Why is this needed?

File: libraries/AP_BattMonitor/AP_BattMonitor_FuelFlow.cpp [hunk 2]
```diff
@@ -0,0 +1,121 @@
+#include <AP_HAL/AP_HAL.h>
+#include <AP_Common/AP_Common.h>
+#include <AP_Math/AP_Math.h>
+#include "AP_BattMonitor.h"
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_FuelFlow.cpp]: This will be included automatically by the AP_BattMonitor_FuelFlow header.

File: libraries/AP_BattMonitor/AP_BattMonitor_FuelFlow.cpp [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    _state.last_time_micros = now_us;
+
+    // map amps to litres/hour
+    _state.current_amps = litres_pec_sec * (60*60);
+
+    // map consumed_mah to consumed millilitres
+    _state.consumed_mah += litres * 1000;
+
+    // map consumed_wh using fixed voltage
+    _state.consumed_wh += litres * 1000 * _state.voltage;
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_FuelFlow.cpp]: If you are going to force `_state.voltage` to 1 then why multiply by it?

File: libraries/AP_BattMonitor/AP_BattMonitor_FuelFlow.cpp [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+    uint32_t now_us = AP_HAL::micros();
+    if (_state.last_time_micros == 0) {
+        // need initial time, so we can work out expected pulse rate
+        _state.last_time_micros = now_us;
+        return;
+    }
+    float dt = (now_us - _state.last_time_micros) * 1.0e-6;
+
+    if (dt < 1 && irq_state.pulse_count == 0) {
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_FuelFlow.cpp]: if dt > 1 amd we haven't gotten healthy is this an indicator that we should be setting healthy to false?

File: libraries/AP_BattMonitor/AP_BattMonitor_FuelFlow.h [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+public:
+
+    /// Constructor
+    AP_BattMonitor_FuelFlow(AP_BattMonitor &mon, AP_BattMonitor::BattMonitor_State &mon_state, AP_BattMonitor_Params &params);
+
+    /// Read the battery voltage and current.  Should be called at 10hz
+    void read() override;
+
+    /// returns true if battery monitor provides consumed energy info
+    bool has_consumed_energy() const override { return has_current(); }
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_FuelFlow.h]: Why not just return true here? This monitor can't change at runtime, it will always have current. (The compiler may or may not be picking this up, but it's less clear by calling the method anyways).

File: libraries/AP_BattMonitor/AP_BattMonitor_FuelFlow.cpp (hunk 2 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_FuelFlow.cpp]: fixed

File: libraries/AP_BattMonitor/AP_BattMonitor_FuelFlow.cpp (hunk 3 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_FuelFlow.cpp]: fixed

File: libraries/AP_BattMonitor/AP_BattMonitor_FuelFlow.h (hunk 5 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_FuelFlow.h]: fixed

File: libraries/AP_BattMonitor/AP_BattMonitor_FuelFlow.cpp (hunk 4 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_FuelFlow.cpp]: no, as the comment states, we don't want to be unhealthy when in pre-arm, with zero throttle

Comment (WickedShell) [reply libraries/AP_BattMonitor/AP_BattMonitor_FuelFlow.cpp]: That makes sense. Do we want to just set the `_state.healthy` in the constructor (or init) then? There isn't any value in doing it each loop for the driver.

Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_FuelFlow.cpp]: done

File: libraries/AP_BattMonitor/AP_BattMonitor_FuelFlow.cpp (hunk 1 above)
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_FuelFlow.cpp]: fixed

**Comment by davidbuzz** [comment]
looks good, but I note that there's no way to zero the "consumed" amount other than a reboot, which isn't always a practical pre-flight solution with the engine already running.   Eg, many fuel-based RC planes use a "taxi tank" or "startup tank" which keeps the in-board tanks full right up to the moment of departure, so it would be helpful to be able to zero  the consumed amount/s.   ( ie _state.consumed_wh  and _state.consumed_mah are never reset back to zero anywhere, maybe a mavlink packet to do it? )

**Comment by tridge** [comment]
@davidbuzz yes, it should be part of the PREFLIGHT_CALIBRATION options
